pkg/healthv2/provider.go

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package healthv2
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/cilium/cilium/pkg/healthv2/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/index"
+	"github.com/sirupsen/logrus"
+)
+
+type HealthProviderParams struct {
+	cell.In
+	Lifecycle cell.Lifecycle
+	DB        *statedb.DB
+}
+
+type HealthProvider struct {
+	db           *statedb.DB
+	stopped      atomic.Bool
+	statusTable  statedb.RWTable[types.Status]
+	Statuses     statedb.Table[types.Status]
+	PrimaryIndex statedb.Index[types.Status, types.HealthID]
+	LevelIndex   statedb.Index[types.Status, types.Level]
+}
+
+const HealthTable = "health"
+
+func newHealthV2Provider(params HealthProviderParams) (types.Provider, *HealthProvider, error) {
+	p := &HealthProvider{
+		db: params.DB,
+	}
+	p.PrimaryIndex = statedb.Index[types.Status, types.HealthID]{
+		Name: "identifier",
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.ID.String()))
```

Comment:
Perhaps nicer/better to use the existing Strigger index helpers https://github.com/cilium/cilium/blob/main/pkg/statedb/index/string.go#L12

Reply:
I think the idea was to have non-terminated key so prefix searches would work. We discussed with @tommyp1ckles about also changing index.String to not terminate and just rely on the iterators to cut off after length mismatches. Though we can also add e.g. index.StringPrefix which is like index.String but no termination and then use that for FromKey, but perhaps less confusing overall if we just drop the key termination.

Reply:
@dylandreimerink @joamaki yeah, I'd like to play around adding support for prefix querys, including partial key match. For example something like:

agent.control-plane.endpoint-manager.endpoint-* that might match a set of endpoint specific status such as:

 ```
agent.control-plane.endpoint-manager.endpoint-1234.k8s-sync
agent.control-plane.endpoint-manager.endpoint-1234.datapath-reconcile
agent.control-plane.endpoint-manager.endpoint-111.k8s-sync
agent.control-plane.endpoint-manager.endpoint-111.datapath-reconcile
...
```